[URGENT BUG] Failed to create PNG

We cant compile our develop branch currently because its failing to create png's

```
2020-07-01 12:22:17.085 ERROR 11260 --- [  restartedMain] o.s.boot.SpringApplication               : Application run failed

java.lang.RuntimeException: Failed to create PNG
	at de.bonndan.nivio.output.map.PNGRenderCache.asByteArray(PNGRenderCache.java:104) ~[classes/:na]
	at de.bonndan.nivio.output.map.PNGRenderCache.getPNG(PNGRenderCache.java:60) ~[classes/:na]
	at de.bonndan.nivio.output.map.PNGRenderCache.onApplicationEvent(PNGRenderCache.java:92) ~[classes/:na]
	at de.bonndan.nivio.output.map.PNGRenderCache.onApplicationEvent(PNGRenderCache.java:35) ~[classes/:na]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:403) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:360) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at de.bonndan.nivio.input.Indexer.reIndex(Indexer.java:62) ~[classes/:na]
	at de.bonndan.nivio.input.IndexEventListener.onApplicationEvent(IndexEventListener.java:27) ~[classes/:na]
	at de.bonndan.nivio.input.IndexEventListener.onApplicationEvent(IndexEventListener.java:13) ~[classes/:na]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:403) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:360) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at de.bonndan.nivio.input.StartupListener.lambda$onApplicationEvent$0(StartupListener.java:55) ~[classes/:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[na:na]
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[na:na]
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[na:na]
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[na:na]
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[na:na]
	at de.bonndan.nivio.input.StartupListener.onApplicationEvent(StartupListener.java:55) ~[classes/:na]
	at de.bonndan.nivio.input.StartupListener.onApplicationEvent(StartupListener.java:24) ~[classes/:na]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:403) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:360) ~[spring-context-5.2.6.RELEASE.jar:5.2.6.RELEASE]
	at org.springframework.boot.context.event.EventPublishingRunListener.running(EventPublishingRunListener.java:103) ~[spring-boot-2.2.7.RELEASE.jar:2.2.7.RELEASE]
	at org.springframework.boot.SpringApplicationRunListeners.running(SpringApplicationRunListeners.java:77) ~[spring-boot-2.2.7.RELEASE.jar:2.2.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:330) ~[spring-boot-2.2.7.RELEASE.jar:2.2.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.2.7.RELEASE.jar:2.2.7.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.2.7.RELEASE.jar:2.2.7.RELEASE]
	at de.bonndan.nivio.Application.main(Application.java:33) ~[classes/:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49) ~[spring-boot-devtools-2.2.7.RELEASE.jar:2.2.7.RELEASE]
Caused by: org.apache.batik.transcoder.TranscoderException: null
Enclosed Exception:
null:0
The URI "http://localhost:8080/vendoricons//http://download.redis.io/logocontest/82.png"
on element <image> can't be opened because:
JDK URL is corrupt or unsupported variant
	at org.apache.batik.transcoder.SVGAbstractTranscoder.transcode(SVGAbstractTranscoder.java:226) ~[batik-transcoder-1.12.jar:na]
	at org.apache.batik.transcoder.image.ImageTranscoder.transcode(ImageTranscoder.java:92) ~[batik-transcoder-1.12.jar:na]
	at org.apache.batik.transcoder.XMLAbstractTranscoder.transcode(XMLAbstractTranscoder.java:142) ~[batik-transcoder-1.12.jar:na]
	at org.apache.batik.transcoder.SVGAbstractTranscoder.transcode(SVGAbstractTranscoder.java:156) ~[batik-transcoder-1.12.jar:na]
	at de.bonndan.nivio.output.map.PNGRenderCache.asByteArray(PNGRenderCache.java:102) ~[classes/:na]
	... 44 common frames omitted
Caused by: org.apache.batik.bridge.BridgeException: null:0
The URI "http://localhost:8080/vendoricons//http://download.redis.io/logocontest/82.png"
on element <image> can't be opened because:
JDK URL is corrupt or unsupported variant
	at org.apache.batik.bridge.UserAgentAdapter.getBrokenLinkDocument(UserAgentAdapter.java:448) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.SVGImageElementBridge.createRasterImageNode(SVGImageElementBridge.java:641) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.SVGImageElementBridge.createImageGraphicsNode(SVGImageElementBridge.java:339) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.SVGImageElementBridge.buildImageGraphicsNode(SVGImageElementBridge.java:179) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.SVGImageElementBridge.createGraphicsNode(SVGImageElementBridge.java:121) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:213) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.GVTBuilder.buildGraphicsNode(GVTBuilder.java:219) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.GVTBuilder.buildComposite(GVTBuilder.java:171) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.bridge.GVTBuilder.build(GVTBuilder.java:82) ~[batik-bridge-1.12.jar:na]
	at org.apache.batik.transcoder.SVGAbstractTranscoder.transcode(SVGAbstractTranscoder.java:208) ~[batik-transcoder-1.12.jar:na]
	... 48 common frames omitted

2020-07-01 12:22:17.089  INFO 11260 --- [  restartedMain] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopping...
2020-07-01 12:22:17.089  INFO 11260 --- [  restartedMain] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=false, SimpleBrokerMessageHandler [DefaultSubscriptionRegistry[cache[0 destination(s)], registry[0 sessions]]]]
2020-07-01 12:22:17.090  INFO 11260 --- [  restartedMain] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopped.
2020-07-01 12:22:17.092  INFO 11260 --- [  restartedMain] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'brokerChannelExecutor'
2020-07-01 12:22:17.092  INFO 11260 --- [  restartedMain] o.s.s.c.ThreadPoolTaskScheduler          : Shutting down ExecutorService 'messageBrokerTaskScheduler'
2020-07-01 12:22:17.093  INFO 11260 --- [  restartedMain] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'clientOutboundChannelExecutor'
2020-07-01 12:22:17.093  INFO 11260 --- [  restartedMain] o.s.s.concurrent.ThreadPoolTaskExecutor  : Shutting down ExecutorService 'clientInboundChannelExecutor'

Process finished with exit code 0
```

--- a/src/main/java/de/bonndan/nivio/input/http/HttpService.java
+++ b/src/main/java/de/bonndan/nivio/input/http/HttpService.java
@@ -24,6 +24,11 @@ public class HttpService {
         return executeRequest(client, new HttpGet(url.toURI()));
     }
 
+    /**
+     * @param url
+     * @return
+     * @throws URISyntaxException
+     */
     public CachedResponse getResponse(URL url) throws URISyntaxException {
 
         HttpGet request = new HttpGet(url.toURI());
@@ -31,10 +36,16 @@ public class HttpService {
         request.setHeader(new BasicHeader("Cache-Control", "no-cache"));
 
         try (CloseableHttpClient client = HttpClients.createDefault(); CloseableHttpResponse response = client.execute(request)) {
-            return new CachedResponse(response.getAllHeaders(), response.getEntity());
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode < 400) {
+                return new CachedResponse(response.getAllHeaders(), response.getEntity());
+            } else {
+                throw new RuntimeException("Failed to fetch " + request.getURI() + ", status: " + statusCode);
+            }
         } catch (IOException ex) {
             throw new RuntimeException("Failed to fetch " + request.getURI(), ex);
         }
+
     }
 
     public String getWithBasicAuth(URL url, String username, String password) throws IOException, AuthenticationException, URISyntaxException {
@@ -58,7 +69,7 @@ public class HttpService {
     private String executeRequest(CloseableHttpClient client, HttpGet request) throws IOException {
         try (CloseableHttpResponse response = client.execute(request)) {
             if (response.getStatusLine().getStatusCode() == 200) {
-                return  EntityUtils.toString(response.getEntity());
+                return EntityUtils.toString(response.getEntity());
             } else {
                 throw new RuntimeException("Got " + response.getStatusLine().getStatusCode() + " while reading");
             }

--- a/src/main/java/de/bonndan/nivio/output/LocalServer.java
+++ b/src/main/java/de/bonndan/nivio/output/LocalServer.java
@@ -16,6 +16,7 @@ import org.springframework.util.StringUtils;
 import java.net.InetAddress;
 import java.net.MalformedURLException;
 import java.net.URL;
+import java.util.Base64;
 
 import static de.bonndan.nivio.output.icons.Icons.DEFAULT_ICON;
 
@@ -39,7 +40,7 @@ public class LocalServer implements EnvironmentAware {
 
     public LocalServer(@Value("${nivio.baseUrl:}") String baseUrl, VendorIcons vendorIcons) {
         if (!StringUtils.isEmpty(baseUrl)) {
-            this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length()-1) : baseUrl;
+            this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
         } else {
             this.baseUrl = "http://" + host() + ":" + port();
         }
@@ -60,6 +61,16 @@ public class LocalServer implements EnvironmentAware {
             LOGGER.warn("Failed to build url for {}", path);
             return defaultIcon;
         }
+    }
+
+    /**
+     * Returns the visible url.
+     *
+     * @param parts path to add, concatenated by "/"
+     * @return url with host, port
+     */
+    public URL getUrl(String... parts) {
+        return getUrl(StringUtils.arrayToDelimitedString(parts, "/"));
     }
 
     public URL getIconUrl(Labeled item) {
@@ -106,19 +117,35 @@ public class LocalServer implements EnvironmentAware {
 
     private URL proxiedUrl(URL url) {
 
-        if (imageProxy == null){
+        if (imageProxy == null) {
             URL imageProxy = getUrl(VENDORICONS_PATH);
             setImageProxy(imageProxy.toString());
         }
 
         if (!StringUtils.isEmpty(imageProxy)) {
             try {
-                return new URL(imageProxy + "//" + url.toString());
+                return new URL(imageProxy + "/" + Base64.getUrlEncoder().encodeToString(url.toString().getBytes()));
             } catch (MalformedURLException e) {
-                LOGGER.error("Failed to build image proxy url from {}", imageProxy + "//" + url.getPath());
+                LOGGER.error("Failed to build image proxy url from {}", imageProxy + "/" + url.getPath());
             }
         }
         return url;
+    }
+
+    /**
+     * Turns the base 64 encode vendor icon url back into the orginal icon url.
+     *
+     * @param requestURI base64 encoded request including nivio path
+     * @return string icon url
+     */
+    public static String deproxyUrl(String requestURI) {
+        String part = VENDORICONS_PATH ;
+        String iconRequestURI = requestURI.split(part)[1];
+        iconRequestURI = StringUtils.trimLeadingCharacter(iconRequestURI, '/');
+
+        iconRequestURI = new String(Base64.getUrlDecoder().decode(iconRequestURI));
+
+        return iconRequestURI;
     }
 
     public void setImageProxy(String imageProxy) {

--- a/src/main/java/de/bonndan/nivio/output/icons/IconsController.java
+++ b/src/main/java/de/bonndan/nivio/output/icons/IconsController.java
@@ -9,7 +9,6 @@ import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.stereotype.Controller;
-import org.springframework.util.StringUtils;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 
@@ -32,7 +31,6 @@ public class IconsController {
     private static final Logger LOGGER = LoggerFactory.getLogger(IconsController.class);
 
 
-
     private final Map<String, CachedResponse> imageCache;
 
     private final HttpService httpService;
@@ -46,9 +44,14 @@ public class IconsController {
     public ResponseEntity<byte[]> icons(HttpServletRequest request) {
 
         String requestURI = request.getRequestURI();
-        String part = VENDORICONS_PATH + "/";
-        String iconRequestURI = requestURI.substring(part.length());
-        iconRequestURI = StringUtils.trimLeadingCharacter(iconRequestURI, '/');
+        String iconRequestURI;
+        try {
+            iconRequestURI = LocalServer.deproxyUrl(requestURI);
+        } catch (IllegalArgumentException ex) {
+            LOGGER.error("Failed to decode " + requestURI + " : " + ex.getMessage());
+            return ResponseEntity.badRequest().build();
+        }
+
         if (imageCache.containsKey(iconRequestURI)) {
             return sendResponse(imageCache.get(iconRequestURI));
         }

--- a/src/main/java/de/bonndan/nivio/output/icons/VendorIcons.java
+++ b/src/main/java/de/bonndan/nivio/output/icons/VendorIcons.java
@@ -29,7 +29,7 @@ public class VendorIcons {
         vendoricons.put("kubernetes", "https://raw.githubusercontent.com/kubernetes/kubernetes/master/logo/logo.png");
 
         //redis.io/topics/trademark
-        vendoricons.put("redis", "http://download.redis.io/logocontest/82.png");
+        vendoricons.put("redis", "https://redis.io/images/redis-white.png");
         vendoricons.put("prometheus", "https://raw.githubusercontent.com/prometheus/docs/master/static/prometheus_logo.png");
     }
 

--- a/src/test/java/de/bonndan/nivio/output/LocalServerTest.java
+++ b/src/test/java/de/bonndan/nivio/output/LocalServerTest.java
@@ -67,7 +67,9 @@ class LocalServerTest {
     public void returnsIcon() {
         Item item = new Item();
         item.setLabel(Label.icon, "http://my.icon");
-        assertTrue(localServer.getIconUrl(item).toString().contains("http://my.icon"));
+        String s = localServer.getIconUrl(item).toString();
+        assertEquals("http://localhost:8080/vendoricons/aHR0cDovL215Lmljb24=",s);
+        assertEquals("http://my.icon", LocalServer.deproxyUrl(s));
     }
 
 
@@ -75,7 +77,9 @@ class LocalServerTest {
     public void usesVendorIcon() {
         Item item = new Item();
         item.setLabel(Label.icon, VENDOR_PREFIX + "redis");
-        assertTrue(localServer.getIconUrl(item).toString().contains("http://download.redis.io/logocontest/82.png"));
+        String s = localServer.getIconUrl(item).toString();
+        assertEquals("http://localhost:8080/vendoricons/aHR0cDovL2Rvd25sb2FkLnJlZGlzLmlvL2xvZ29jb250ZXN0LzgyLnBuZw==", s);
+        assertEquals("http://download.redis.io/logocontest/82.png", LocalServer.deproxyUrl(s));
     }
 
     @Test
@@ -88,7 +92,7 @@ class LocalServerTest {
 
         Item item = new Item();
         item.setLabel(Label.icon, VENDOR_PREFIX + "redis");
-        assertEquals(urlprefix + "//" + "http://download.redis.io/logocontest/82.png", localServer.getIconUrl(item).toString());
+        assertEquals(urlprefix + "/aHR0cDovL2Rvd25sb2FkLnJlZGlzLmlvL2xvZ29jb250ZXN0LzgyLnBuZw==", localServer.getIconUrl(item).toString());
     }
 
     @Test
@@ -102,7 +106,7 @@ class LocalServerTest {
         Item item = new Item();
         item.setLabel(Label.icon, "http://my.icon");
         String urlprefix = String.format("http://localhost:%d", wireMockServer.port());
-        assertEquals(urlprefix + "//" + "http://my.icon", localServer.getIconUrl(item).toString());
+        assertEquals(urlprefix + "/aHR0cDovL215Lmljb24=", localServer.getIconUrl(item).toString());
 
     }
 

